### PR TITLE
scan barcode

### PR DESCRIPTION
**Pull request by at-haytran**
## Description
A few sentences describing the overall goals of the pull request's commits.
## Reference Requirements
- Task ID: TI-199

## Reference Isuees
#{issue_id}

## Migrations
YES

## Impacted Areas in Application
List general components of the application that this PR will affect:

-

## SQL query and `explain`
List of all queries and attach the result of each explaining query (in text, screenshot...)

- Refer to [Using Explain with MySQL](https://viblo.asia/p/su-dung-explain-de-toi-uu-cau-lenh-mysql-BYjv44gmvxpV)

### CODE REVIEW

File: resources/views/admin/books/index.blade.php [hunk 1]
```diff
@@ -5,6 +5,12 @@
...
 <div class="content-wrapper">
+    <section class="container">
+        <div id="scanner-container"></div>
+        <input type="button" id="btn" value="Start/Stop the scanner"/>
+        <div class="alert alert-danger alert-dismissible fade in" hidden>
+        </div>
+    </section>
     <!-- Content Header (Page header) -->
     <section class="content-header">
         <h2>{{trans('book.list_book')}}</h2>
```
Comment (at-haphan) [inline resources/views/admin/books/index.blade.php]: dấu cách 2 đầu

File: public/js/readBarcode.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            }
+        },
+
+    }, function (err) {
+        if (err) {
+            console.log(err);
+            return
+        }
+
+        console.log("Initialization finished. Ready to start");
```
Comment (at-quangtran) [inline public/js/readBarcode.js]: console.log

File: routes/web.php [hunk 3]
```diff
@@ -31,6 +31,7 @@
     Route::get('/', 'HomeController@index');
     Route::resource('users', 'UserController');
     Route::resource('books', 'BookController');
+    Route::post('books/readBarcode', 'BookController@readBarcode')->name('readBarcode');
```
Comment (at-quangtran) [inline routes/web.php]: read-barcode

Comment (at-haytran) [reply routes/web.php]: cái đó cho route name hay sao anh.

File: public/js/readBarcode.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                    showTransformed: true,
+                    showTransformedBox: true,
+                    showBB: true
+                }
+            }
+        },
+
+    }, function (err) {
+        if (err) {
+            console.log(err);
```
Comment (at-quangtran) [inline public/js/readBarcode.js]: console

File: public/js/readBarcode.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+            if (result.codeResult && result.codeResult.code) {
+                Quagga.ImageDebug.drawPath(result.line, { x: 'x', y: 'y' }, drawingCtx, { color: 'red', lineWidth: 3 });
+            }
+        }
+    });
+
+
+    Quagga.onDetected(function (result) {
+        console.log("Barcode detected and processed : [" + result.codeResult.code + "]", result);
```
Comment (at-quangtran) [inline public/js/readBarcode.js]: console

File: public/js/readBarcode.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        return Object.keys(counts).sort(function(curKey,nextKey) {
+            return counts[curKey] < counts[nextKey];
+    });
+}
+
+function startScanner() {
+    Quagga.onDetected(function(result) {
+        var last_code = result.codeResult.code;
+        code = last_code;
+        console.log(code);
```
Comment (at-quangtran) [inline public/js/readBarcode.js]: đây nữa này

File: database/migrations/2018_07_12_014352_update2_books_table.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+use Illuminate\Support\Facades\Schema;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Database\Migrations\Migration;
+
+class Update2BooksTable extends Migration
```
Comment (at-quangtran) [inline database/migrations/2018_07_12_014352_update2_books_table.php]: tên file, class nên đặt ý nghĩa gợi nhớ nội dung 
2018_07_12_014352_update2_books_table.php => 2018_07_12_014352_add_column_barcode_in_books_table.php 
Update2BooksTable => AddColumnBarcodeInBooksTable

**Pull request merged**